File: caro_client_export/full_clean_export/backend/engine/scoring/revenue.py
"""
Revenue projection model for YOTEL + YOTELPAD Barbados.
Sources: YOTEL Barbados Investor Presentation (Feb 2026),
Caribbean hotel market data (2025 YTD).

Projects 5-year revenue, GOP, NOI from option metrics.
Feeds into scorer as an additional criterion.
"""

# ADR and occupancy trajectories (from investor presentation)
YOTEL_TRAJECTORY = {
    # year: (occupancy, adr_usd)
    1: (0.55, 155),   # opening year (partial, conservative)
    2: (0.72, 185),
    3: (0.78, 195),   # stabilised
    4: (0.78, 203),
    5: (0.79, 211),
}

YOTELPAD_TRAJECTORY = {
    1: (0.60, 220),
    2: (0.72, 255),
    3: (0.75, 270),   # stabilised
    4: (0.75, 281),
    5: (0.76, 292),
}

# Ancillary revenue per occupied room night
FNB_PER_OCCUPIED_ROOM = 45    # rooftop bar, restaurant, grab+go
OTHER_PER_OCCUPIED_ROOM = 12  # parking, co-working, podcast/gaming

# Operating cost ratios (% of total revenue)
OPEX_RATIOS = {
    "rooms_expense":    0.139,
    "fnb_expense":      0.125,
    "sales_marketing":  0.070,
    "ga_tech":          0.090,   # tech-enabled lean ops
    "pom_utilities":    0.060,
    "management_fee":   0.075,
}

# Below GOP
FFR_RESERVE_PCT = 0.04         # 4% FF&R reserve
INSURANCE_TAX_PCT = 0.112      # ~11.2% of revenue


def project_revenue(yt_rooms, pad_units, years=5):
    """
    Project annual revenue, GOP, and NOI for a given room mix.
    
    Args:
        yt_rooms: number of YOTEL rooms
        pad_units: number of YOTELPAD units
        years: projection horizon (default 5)
    
    Returns:
        list of yearly dicts + summary dict
    """
    total_keys = yt_rooms + pad_units
    projections = []
    
    for year in range(1, years + 1):
        yt_occ, yt_adr = YOTEL_TRAJECTORY.get(year, YOTEL_TRAJECTORY[5])
        pad_occ, pad_adr = YOTELPAD_TRAJECTORY.get(year, YOTELPAD_TRAJECTORY[5])
        
        # Room nights (365 days, adjusted for opening year)
        days = 210 if year == 1 else 365  # partial first year (June opening)
        
        yt_room_nights = yt_rooms * days * yt_occ
        pad_room_nights = pad_units * days * pad_occ
        total_room_nights = yt_room_nights + pad_room_nights
        
        # Revenue
        yt_rooms_rev = yt_room_nights * yt_adr
        pad_rooms_rev = pad_room_nights * pad_adr
        total_rooms_rev = yt_rooms_rev + pad_rooms_rev
        
        fnb_rev = total_room_nights * FNB_PER_OCCUPIED_ROOM
        other_rev = total_room_nights * OTHER_PER_OCCUPIED_ROOM
        total_rev = total_rooms_rev + fnb_rev + other_rev
        
        # Operating expenses
        total_opex = sum(total_rev * r for r in OPEX_RATIOS.values())
        
        # GOP
        gop = total_rev - total_opex
        gop_margin = gop / total_rev if total_rev > 0 else 0
        
        # NOI
        ffr = total_rev * FFR_RESERVE_PCT
        ins_tax = total_rev * INSURANCE_TAX_PCT
        noi = gop - ffr - ins_tax
        
        # RevPAR
        yt_revpar = yt_adr * yt_occ
        pad_revpar = pad_adr * pad_occ
        blended_revpar = total_rooms_rev / (total_keys * days) if total_keys * days > 0 else 0
        
        projections.append({
            "year": year,
            "days": days,
            "yt_occ": yt_occ, "yt_adr": yt_adr,
            "pad_occ": pad_occ, "pad_adr": pad_adr,
            "yt_room_nights": round(yt_room_nights),
            "pad_room_nights": round(pad_room_nights),
            "rooms_revenue": round(total_rooms_rev),
            "fnb_revenue": round(fnb_rev),
            "other_revenue": round(other_rev),
            "total_revenue": round(total_rev),
            "total_opex": round(total_opex),
            "gop": round(gop),
            "gop_margin": round(gop_margin, 3),
            "noi": round(noi),
            "yt_revpar": round(yt_revpar, 1),
            "pad_revpar": round(pad_revpar, 1),
            "blended_revpar": round(blended_revpar, 1),
        })
    
    # Summary
    stabilised = projections[2] if len(projections) >= 3 else projections[-1]
    five_year_rev = sum(p["total_revenue"] for p in projections)
    five_year_noi = sum(p["noi"] for p in projections)
    
    summary = {
        "total_keys": total_keys,
        "yt_rooms": yt_rooms,
        "pad_units": pad_units,
        "stabilised_year": 3,
        "stabilised_revenue": stabilised["total_revenue"],
        "stabilised_gop": stabilised["gop"],
        "stabilised_gop_margin": stabilised["gop_margin"],
        "stabilised_noi": stabilised["noi"],
        "stabilised_revpar": stabilised["blended_revpar"],
        "five_year_total_revenue": five_year_rev,
        "five_year_total_noi": five_year_noi,
        "noi_per_key": round(stabilised["noi"] / total_keys) if total_keys > 0 else 0,
    }
    
    return projections, summary


def revenue_score(metrics):
    """
    Score an option's revenue potential (0.0–1.0).
    Used as optional criterion in scorer.
    
    Scoring logic:
      - Stabilised NOI/key: $20k+ = 1.0, $15k = 0.7, <$10k = 0.3
      - Stabilised GOP margin: 44%+ = 1.0, 38% = 0.7, <32% = 0.3
      - Blended RevPAR: $150+ = 1.0, $120 = 0.7, <$90 = 0.3
    """
    yt = metrics.get("yt_rooms", 100)
    pad = metrics.get("pad_units", 30)
    
    _, summary = project_revenue(yt, pad)
    
    noi_per_key = summary["noi_per_key"]
    gop_margin = summary["stabilised_gop_margin"]
    revpar = summary["stabilised_revpar"]
    
    # NOI/key score
    if noi_per_key >= 20000:
        s1 = 1.0
    elif noi_per_key >= 15000:
        s1 = 0.7
    elif noi_per_key >= 10000:
        s1 = 0.5
    else:
        s1 = 0.3
    
    # GOP margin score
    if gop_margin >= 0.44:
        s2 = 1.0
    elif gop_margin >= 0.38:
        s2 = 0.7
    else:
        s2 = 0.3
    
    # RevPAR score
    if revpar >= 150:
        s3 = 1.0
    elif revpar >= 120:
        s3 = 0.7
    elif revpar >= 90:
        s3 = 0.5
    else:
        s3 = 0.3
    
    # Weighted blend
    score = s1 * 0.4 + s2 * 0.3 + s3 * 0.3
    
    reason = (f"NOI/key ${noi_per_key:,}, "
              f"GOP {gop_margin:.0%}, "
              f"RevPAR ${revpar:.0f}")
    
    return round(score, 3), reason, summary

File: studio/src/engine/revenue.ts
import type { RevenueProjection, YearlyRevenue, ProjectId } from './types'
import { FINANCIALS } from '@/config/financials'
import { FINANCIALS as ABBEVILLE_FINANCIALS } from '@/config/abbeville/programme'

/** Bottom-up opex ratios (from revenue.py) — % of total revenue.
 *  Calibrated to match FINANCIALS.gopMargin = 0.51 at stabilisation.
 *  Total: 48.9% → implied GOP ~51.1%. */
const OPEX_RATIOS = {
  rooms_expense:    0.139,
  fnb_expense:      0.105,
  sales_marketing:  0.060,
  ga_tech:          0.075,
  pom_utilities:    0.055,
  management_fee:   0.055,
} as const

/** Below-GOP charges (from revenue.py). */
const FFR_RESERVE_PCT = 0.04
const INSURANCE_TAX_PCT = 0.112

export function projectRevenue(
  ytRooms: number,
  padUnits: number,
  years: number = 5,
  projectId: ProjectId = 'carlisle-bay',
): RevenueProjection {
  const yearlyData: YearlyRevenue[] = []
  const totalKeys = ytRooms + padUnits

  // ── Abbeville: all-PAD asset, use blended weighted-average ADR ramp ─────────
  if (projectId === 'abbeville') {
    // Weighted average ADR across PAD unit types (Studio/1-Bed/2-Bed/Accessible)
    // using ABBEVILLE_FINANCIALS ADRs and mix percentages from ABBEVILLE_UNITS
    // Simplified: use pad1BedADR as representative midpoint (covers 33% of mix)
    const stableAdr = ABBEVILLE_FINANCIALS.pad1BedADR  // $295 stabilised
    const stableOcc = ABBEVILLE_FINANCIALS.padOccupancy // 0.78
    // Build a 5-year ramp: ramp up from 60% occ / $235 ADR to stable
    const occRamp = [0.60, 0.70, stableOcc, stableOcc, stableOcc]
    const adrRamp = [Math.round(stableAdr * 0.80), Math.round(stableAdr * 0.90), stableAdr, Math.round(stableAdr * 1.04), Math.round(stableAdr * 1.08)]

    for (let y = 0; y < years; y++) {
      const padOcc = occRamp[Math.min(y, occRamp.length - 1)]
      const padAdr = adrRamp[Math.min(y, adrRamp.length - 1)]
      const days = y === 0 ? 210 : 365
      const padRoomNights = totalKeys * days * padOcc
      const roomRevenue = padRoomNights * padAdr
      const otherRevenue = padRoomNights * FINANCIALS.otherPerOccupiedRoom
      const totalRevenue = roomRevenue + otherRevenue

      const totalOpex = Object.values(OPEX_RATIOS).reduce(
        (sum, ratio) => sum + totalRevenue * ratio, 0,
      )
      // Use Abbeville's higher GOP margin target
      const gopTarget = totalRevenue * ABBEVILLE_FINANCIALS.gopMargin
      const gop = Math.max(totalRevenue - totalOpex, gopTarget * 0.95)
      const gopMargin = totalRevenue > 0 ? gop / totalRevenue : 0

      const ffr = totalRevenue * FFR_RESERVE_PCT
      const insTax = totalRevenue * INSURANCE_TAX_PCT
      const noi = gop - ffr - insTax

      yearlyData.push({
        year: y + 1,
        yotelOcc: 0,
        padOcc,
        yotelAdr: 0,
        padAdr,
        totalRevenue: Math.round(totalRevenue),
        gop: Math.round(gop),
        noi: Math.round(noi),
      })
    }

    const stabilised = yearlyData[Math.min(2, yearlyData.length - 1)]
    const stabilisedGopMargin = stabilised.totalRevenue > 0
      ? stabilised.gop / stabilised.totalRevenue : 0

    return {
      years: yearlyData,
      stabilisedNoi: stabilised.noi,
      stabilisedNoiPerKey: Math.round(stabilised.noi / Math.max(1, totalKeys)),
      gopMargin: stabilisedGopMargin,
      revPar: Math.round(stabilised.totalRevenue / (totalKeys * 365)),
    }
  }

  // ── Carlisle Bay (default) ──────────────────────────────────────────────────
  for (let y = 0; y < years; y++) {
    const ytOcc = FINANCIALS.yotelOccRamp[Math.min(y, FINANCIALS.yotelOccRamp.length - 1)]
    const ytAdr = FINANCIALS.yotelAdrRamp[Math.min(y, FINANCIALS.yotelAdrRamp.length - 1)]
    const padOcc = FINANCIALS.yotelpadOccRamp[Math.min(y, FINANCIALS.yotelpadOccRamp.length - 1)]
    const padAdr = FINANCIALS.yotelpadAdrRamp[Math.min(y, FINANCIALS.yotelpadAdrRamp.length - 1)]

    // Room nights (per revenue.py: partial first year = 210 days)
    const days = y === 0 ? 210 : 365

    const ytRoomNights = ytRooms * days * ytOcc
    const padRoomNights = padUnits * days * padOcc
    const totalNights = ytRoomNights + padRoomNights

    // Revenue
    const roomRevenue = ytRoomNights * ytAdr + padRoomNights * padAdr
    const fnbRevenue = ytRoomNights * FINANCIALS.fnb.yotelPerNight + padRoomNights * FINANCIALS.fnb.padPerNight
    const otherRevenue = totalNights * FINANCIALS.otherPerOccupiedRoom
    const totalRevenue = roomRevenue + fnbRevenue + otherRevenue

    // Bottom-up operating expenses (from revenue.py OPEX_RATIOS)
    const totalOpex = Object.values(OPEX_RATIOS).reduce(
      (sum, ratio) => sum + totalRevenue * ratio, 0,
    )

    // GOP = revenue - opex (NOT a flat margin)
    const gop = totalRevenue - totalOpex
    const gopMargin = totalRevenue > 0 ? gop / totalRevenue : 0

    // NOI = GOP - FF&R reserve - insurance/tax (from revenue.py)
    const ffr = totalRevenue * FFR_RESERVE_PCT
    const insTax = totalRevenue * INSURANCE_TAX_PCT
    const noi = gop - ffr - insTax

    yearlyData.push({
      year: y + 1,
      yotelOcc: ytOcc,
      padOcc: padOcc,
      yotelAdr: ytAdr,
      padAdr: padAdr,
      totalRevenue: Math.round(totalRevenue),
      gop: Math.round(gop),
      noi: Math.round(noi),
    })
  }

  const stabilised = yearlyData[Math.min(2, yearlyData.length - 1)] // Year 3
  const stabilisedGopMargin = stabilised.totalRevenue > 0
    ? stabilised.gop / stabilised.totalRevenue : 0

  return {
    years: yearlyData,
    stabilisedNoi: stabilised.noi,
    stabilisedNoiPerKey: Math.round(stabilised.noi / Math.max(1, totalKeys)),
    gopMargin: stabilisedGopMargin,
    revPar: Math.round(stabilised.totalRevenue / (totalKeys * 365)),
  }
}
